simulator.py

```diff
@@ -0,0 +1,155 @@
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        today = self.__portfolio[-1]
+        cost = 0
+        trading_value = 0
+        profit = 0
+        for instrument, proportion in today.items():
+            if target_data[instrument][-1].sell is not None:
+                now = self.quantity(
+                    proportion,
+                    proportion / target_data[instrument][-1].sell
+                )
+                trading_value += now.value
+                profit += (
+                    (self.cur_asset[instrument].quantity - now.quantity) *
+                    target_data[instrument][-1].sell)
+                cost += cost_function(
+                    self.cur_asset[instrument],
+                    now,
+                    target_data[instrument][-1]
+                )
+                self.cur_asset[instrument] = now
+        basic_value = 1 - trading_value
+        profit += self.cur_asset[region].value - basic_value
+        self.cur_asset[region] = self.quantity(basic_value, basic_value)
+        self.__pnl.append((today.date, profit - cost, cost))
+
+
+class Simulator(SimulateMethod):
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data):
+        SimulateMethod.__init__(
+            self,
+            settings.target[1],
+            list(data['target_data'].keys()))
+        self.alpha_formula = None
+        self.data = data['simulate_data']
+        self.settings = settings
+        self.target_data = data['target_data']
+        self.__key = random.randint(0, 2147483647)
+        self.workdays = data['workdays']
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data.values():
+            dataset.next_date(self.__key)
+        self.target_data.next_date(self.__key)
+        self.workdays.next_date(auth_key=self.__key)
+
+    def run(self):
+        """
+        Carried out simulating process.
+        """
+        self.setting()
+        self.alpha_formula = self.settings.alpha(
+            None, self.data, self.settings.parameters)
```

Comment:
The first argument shouldn't be `None`. If its value is always `None`, we shouldn't design this argument from the very start.
While this argument is named `date` in our template, its real meaning is startdate (which is set in AlphaSettings), and users can use this date to initialize their alphas.